pyispyb/core/modules/samples.py

```diff
@@ -238,24 +245,204 @@ def get_samples(
+    crystal: models.Crystal = None
+    if isinstance(sample.Crystal, schema.SampleCrystalUpdate):
+        crystal = with_authorization_proposal(
+            db.session.query(models.Crystal)
+            .filter(models.Crystal.crystalId == sample.Crystal.crystalId)
+            .join(models.Protein)
+            .join(models.Proposal)
+        ).first()
+        if crystal is None:
+            raise HTTPException(
+                status_code=422,
+                detail=f"Could not find Crystal with id {sample.Crystal.crystalId}",
+            )
+        update_model(crystal, sample.Crystal.dict(exclude_unset=True), nested=False)
+    else:
+        # Create new crystal
+        protein = with_authorization_proposal(
+            db.session.query(models.Protein)
+            .filter(models.Protein.proteinId == sample.Crystal.Protein.proteinId)
+            .join(models.Proposal)
+        ).first()
+        if protein is None:
+            raise HTTPException(
+                status_code=422,
+                detail=f"Could not find protein with id {sample.Crystal.Protein.proteinId}",
+            )
+        crystal = models.Crystal(
+            **{**sample.Crystal.dict(), "Protein": protein, "crystal_compositions": []}
+        )
+
+    proposal = crystal.Protein.Proposal
+
+    crystal.crystal_compositions = build_compositions(
+        models.CrystalComposition, sample.Crystal.crystal_compositions, proposal
+    )
+
+    return crystal
+
+
 def create_sample(sample: schema.SampleCreate) -> models.BLSample:
     sample_dict = sample.dict()
-    sample = models.BLSample(**sample_dict)
-    db.session.add(sample)
+    crystal = build_crystal(sample)
+
+    proposal = crystal.Protein.Proposal
+
+    authorize_for_proposal(proposal.proposalId)
+
+    sample_compositions = build_compositions(
+        models.SampleComposition, sample.sample_compositions, proposal
+    )
+
+    new_sample = models.BLSample(
+        **{
+            **sample_dict,
+            "Crystal": crystal,
+            "sample_compositions": sample_compositions,
+        }
+    )
+
+    db.session.add(new_sample)
     db.session.commit()
 
-    new_sample = get_samples(sampleId=sample.sampleId, skip=0, limit=1)
+    new_sample = get_samples(blSampleId=new_sample.blSampleId, skip=0, limit=1)
     return new_sample.first
 
 
-def update_sample(sampleId: int, sample: schema.SampleCreate) -> models.BLSample:
+def update_sample(blSampleId: int, sample: schema.SampleUpdate) -> models.BLSample:
     sample_dict = sample.dict(exclude_unset=True)
-    new_sample = get_samples(sampleId=sampleId, skip=0, limit=1).first
+    old_sample = get_samples(blSampleId=blSampleId, skip=0, limit=1).first
+    update_model(old_sample, sample_dict, nested=False)
+
+    crystal = build_crystal(sample)
+    old_sample.Crystal = crystal
+
+    proposal = crystal.Protein.Proposal
+    authorize_for_proposal(proposal.proposalId)
```

Comment:
I dont think you should need this.
```
old_sample = get_samples(blSampleId=blSampleId, skip=0, limit=1).first
```
will check the authorization of the current user vs the sample and return an empty list if they dont have access, thus throwing IndexError which the route function will catch